functionalDB/chapter.md

```diff
@@ -0,0 +1,1138 @@
+
+For example, let’s look at at the index sketched in Figure 2:
+* the first level stores entity-ids (the blue-ish area) 
+* the second level stores the related attribute-names (the green-ish area)
+* the third level stores the related value (the pink-ish area)
+
+This index is named EAVT, as the top level map holds (E) entity ids, the second level holds (A) attribute names, and the leaves hold (V) values. The (T) comes from the fact that each layer in the database has its own indexes, hence the index itself is relevant for a specific (T) time. 
+
+![image alt text](image_1.png)
+
+Figure 2
+
+Figure 3 shows an index that would be called AVET since:
+
+* First level map holds attribute-name
+* Second level map holds the values (of the attributes)
+* Third level set holds the entity-ids (of the entities whose attribute is at the first level) 
+
+![image alt text](image_2.png)
+
+Figure 3
+
+Our indexes are implemented as a map of maps, where the keys of the root map act as the first level, each such key points to a map whose keys act as the index’s second-level and the values are the index’s third level. Each element in the third level is a set, holding the leaves of the index.
+
+Each index stores the components of a datom as some permutation of its canonical 'EAV' ordering (entity_id, attribute-name, attribute-value). However, when we are working with datoms _outside_ of the index, we expect them to be in canonical format. We thus provide each index with functions *from-eav* and *to-eav* to convert to and from these orderings.
+
+In most database systems, indexes are an optional component; for example, in an RDMBS like postgresql or mysql, you will choose to add indexes only to certain columns in a table. We provide each index with a *usage-pred* function that determines for an attribute whether it should be included in this index or not. 
+
+````clojure
+(defn make-index [from-eav to-eav usage-pred]
+    (with-meta {} {:from-eav from-eav :to-eav to-eav :usage-pred usage-pred}))
+ 
+ (defn from-eav [index] (:from-eav (meta index)))
+ (defn to-eav [index] (:to-eav (meta index)))
+ (defn usage-pred [index] (:usage-pred (meta index)))
+````
+
+In our database there are four indexes - EAVT (as depicted in Figure 2), AVET (as can be seen in Figure 3), VEAT and VAET. We can access these as a vector of values returned from the *indexes* function.
+
+````clojure
+(defn indexes[] [:VAET :AVET :VEAT :EAVT])
+````
+
+To see how all of this comes together, the result of indexing the following five entities is visualized in the table below (the color coding follows the color coding of Figure 2 and Figure 3)

```

Comment:
Great! I was just about to add a comment asking for an example to make the description more concrete.